### PR TITLE
[Merged by Bors] - feat(RingTheory/TwosidedIdeal/Basic): define two-sided-ideal

### PR DESCRIPTION
**Pull request by jjaassoonn**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by j-loreaux** [comment]
How do we feel about having `abbrev TwoSidedIdeal := RingCon`? I don't love the spelling `RingCon`.

**Comment by jjaassoonn** [comment]
> How do we feel about having `abbrev TwoSidedIdeal := RingCon`? I don't love the spelling `RingCon`.

I have added the abbreviation `TwoSidedIdeal`.

File: Mathlib/RingTheory/TwoSidedIdeal/Basic.lean [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+
+open MulOpposite
+
+section definitions
+
+/--
+A two-sided ideal of a ring `R` is a subset of `R` that contains `0` and is closed under addition,
+negation, and absorbs multiplication on both sides.
+-/
+abbrev TwoSidedIdeal (R : Type*) [NonUnitalNonAssocRing R] := RingCon R
```
Comment (jjaassoonn) [inline Mathlib/RingTheory/TwoSidedIdeal/Basic.lean:33]: Should it be `abbrev` or `def`, do we want Lean to unfold `TwoSidedIdeal`?

Comment (j-loreaux) [reply Mathlib/RingTheory/TwoSidedIdeal/Basic.lean:33]: Definitely `abbrev` if we're going to have it at all, but I'll wait for other reviewers to comment about whether we actually want the alternate spelling.

Comment (eric-wieser) [reply Mathlib/RingTheory/TwoSidedIdeal/Basic.lean:33]: I think having this alternate spelling is probably confusing; either we should have one name that functions both as a relation and a set, or we define
```lean
structure TwoSidedIdeal where
  toRingCon : RingCon R
```
to draw a clear distinction between the two APIs.

Comment (jjaassoonn) [reply Mathlib/RingTheory/TwoSidedIdeal/Basic.lean:33]: I am okay both ways, not to use a structure with a single field could save some boilerplate probably. But I don’t really mind boilerplate code.

Comment (jjaassoonn) [reply Mathlib/RingTheory/TwoSidedIdeal/Basic.lean]: Okay. I have switched to one-field-structure approach.

File: Mathlib/RingTheory/TwoSidedIdeal/Basic.lean [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+- a proof that `x * y ∈ S` if `y ∈ S`;
+- a proof that `x * y ∈ S` if `x ∈ S`.
+-/
+def mk' (carrier : Set R)
+    (zero_mem : 0 ∈ carrier)
+    (add_mem : ∀ {x y}, x ∈ carrier → y ∈ carrier → x + y ∈ carrier)
+    (neg_mem : ∀ {x}, x ∈ carrier → -x ∈ carrier)
+    (mul_mem_left : ∀ {x y}, y ∈ carrier → x * y ∈ carrier)
+    (mul_mem_right : ∀ {x y}, x ∈ carrier → x * y ∈ carrier) : TwoSidedIdeal R where
+  ringCon :=
```
Comment (jjaassoonn) [inline Mathlib/RingTheory/TwoSidedIdeal/Basic.lean:118]: Since we are using a structure, might as well provide an alternative constructor

**Review by kbuzzard — APPROVED** [review]
This now looks fine to me.

**Comment by j-loreaux** [comment]
@eric-wieser and @kbuzzard : sorry I didn't get to this sooner, I've been a bit busy. My issue with `structure` has to do with quotients. One of the primary reasons we need two-sided ideals in this generality is to properly talk about quotients of (non-commutative and non-unital) rings by ideals.

With the `abbrev` approach, my thinking is that this would come mostly for free, but I'm concerned that using a `structure` will significantly complicate this.

Actually, this brings me to my second concern, which is not necessarily directed at this PR: how do we plan to deal with quotients anyway? Currently, we make rather involved use of the fact that we have very convenient defeqs between quotients of `AddGroup`s and quotients of `CommRing`s. We're going to have to lose this at some point (I think). Do you have any good ideas for dealing with this?

**Comment by eric-wieser** [comment]
> We're going to have to lose this at some point (I think).

I'm not convinced; I somewhat expect that we will be able to keep it. Quotients by ideals are already defined in terms of `RingCon`, without any issue.

**Comment by j-loreaux** [comment]
In that case I'll just remove my objection / concern. Thanks!

**Comment by eric-wieser** [comment]
> With the `abbrev` approach, my thinking is that this would come mostly for free, but I'm concerned that using a `structure` will significantly complicate this.

What this *does* apply to is the lattice structure, but I think this is just a pile of boilerplate, and not a significant complication.

**Comment by jjaassoonn** [comment]
Can I gently nudge this PR in the direction of merging? perhaps @eric-wieser

**Comment by kbuzzard** [comment]
maintainer merge

File: Mathlib/RingTheory/TwoSidedIdeal/Basic.lean [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+import Mathlib.RingTheory.Congruence.Basic
+
+/-!
+# Two Sided Ideals
+
+In this file, for any `Ring R`, we reinterpret `I : RingCon R` as a two-sided-ideal of a ring.
+
+## Main definitions and results
+
+* `TwoSidedIdeal`: For any `NonUnitalNonAssocRing R`, `TwoSidedIdeal R` is the exactly `RingCon R`.
```
Comment (jcommelin) [inline Mathlib/RingTheory/TwoSidedIdeal/Basic.lean:18]:
```suggestion
* `TwoSidedIdeal`: For any `NonUnitalNonAssocRing R`, `TwoSidedIdeal R` is a wrapper around `RingCon R`.
```

File: Mathlib/RingTheory/TwoSidedIdeal/Basic.lean [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+* `TwoSidedIdeal.setLike`: Every `I : TwoSidedIdeal R` can be interpreted as a set of `R` where
+  `x ∈ I` if and only if `I x 0`.
```
Comment (jcommelin) [inline Mathlib/RingTheory/TwoSidedIdeal/Basic.lean:20]:
```suggestion
  `x ∈ I` if and only if `I.ringCon x 0`.
```

File: Mathlib/RingTheory/TwoSidedIdeal/Basic.lean [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+    (mul_mem_right : ∀ {x y}, x ∈ carrier → x * y ∈ carrier)
+    (x : R) :
+    x ∈ mk' carrier zero_mem add_mem neg_mem mul_mem_left mul_mem_right ↔ x ∈ carrier := by
+  rw [mem_iff]
+  simp [mk']
+
+instance : AddSubgroupClass (TwoSidedIdeal R) R where
+  zero_mem := zero_mem
+  add_mem := @add_mem _ _
+  neg_mem := @neg_mem _ _
```
Comment (jcommelin) [inline Mathlib/RingTheory/TwoSidedIdeal/Basic.lean:148]: I think you can move this up, just after the `neg_mem` lemma.
Then you can use it in theproofs of `sub_mem` and `nsmul_mem` etc...
Of course the current proofs are also one-liners, so this isn't very important.

**Review by jcommelin — COMMENTED** [review]
bors d+

**Pull request closed without merging**